### ISSUE TITLE
Error in predict.ranger.forest(forest, data, predict.all, seed, num.threads,  :    Missing values in data, variable x appears to be constant within groups.. when predicting using undersampled data

### ISSUE THREAD

**Issue by mhmdshari** [mlr-org/mlr#1259]
I use the following code to  undersample the majority class in training data and predict. 

```
  trainTask = makeClassifTask(data = train,target = "cls", positive = "yes")
  testTask = makeClassifTask(data = test, target = "cls",  positive = "yes")
  trainTask = undersample(trainTask, rate = 0.1)
  trainTask =removeConstantFeatures(trainTask)
  Clslearn = makeLearner(clsnam, predict.type = "prob")
  trained = train(Clslearn, trainTask)
  predicted = predict(trained, testTask)
```

For most of the classifiers such as xgboost, ada,avNNet,C50, gbm,IBk, J48 etc I dont have any issues.
But for classif.ranger I get 

```
Error in predict.ranger.forest(forest, data, predict.all, seed, num.threads,  : 
  Missing values in data.
```

and for classif.lda I get  the error
`variable 49 appears to be constant within groups`

Pls note that the above specified algorithms work fine when applied on the original data.

Please kindly help me  to fix these issues.

**Comment by berndbischl** [comment]
we really need reproducible examples in such cases. otherwise nobody can help you.

furthermore:

1) it is very likely that this has nothing to do with mlr, but with the underlying learners.  
lda already tells you what problem it has. and for ranger there have been quite a few bugfixes, very often reported from our side / mlr usage. make sure to use the latest version. 
does the normal classif.randomForest work?

2) why do you create 2 tasks? it would be much easier to create one and use index sets.

**Comment by tjvananne** [comment]
Hello, 

This same error occurred while I was using ranger and I believe I have pinpointed one possible reason why that might be. If you are using the 'respect.unordered.factors = "order"' functionality, then factors are "recoded" for your  training data (based on what I could see in the code on github). 

So, if you have any values of factors in your test data that do not appear in your training data (regardless of if they were combined when you created the factors originally) then you will definitely be presented with this error.

    x1 <- as.factor(c("a", "b", "c", "d", "e", "f", "g"))
    x2 <- c(1, 2, 3, 4, 5, 6, 7)
    x3 <- c(100, 200, 300, 400, 500, 600, 700)
    
    df1 <- data.frame(x1, x2, x3)
    
    str(df1)
    
    set.seed(1776)  # setting seed just because
    train_inx <- sample(1:nrow(df1), size = floor(nrow(df1) * .7))
    
    
    
    train <- df1[train_inx,]
    test <- df1[-train_inx,]
    
    # even though all factor levels are in here, we will get error on prediction
    train$x1  
    
    ranger1 <- ranger(x3 ~.,
                data = train,
                write.forest = T,
                
                # try changing this to "ignore" and there will be no error
                respect.unordered.factors = "order" 
                )
    
    # error
    preds <- predict(ranger1, test)


I haven't read up on exactly when / why to use the "order" flag in the model, but came across this error when randomly generating parameters. I guess "order" should not be used on data sets that have factors with high counts of levels?

**Comment by larskotthoff** [comment]
Thanks for having a look at this. I don't get an error with the code you've posted. Are you using the latest version of ranger?

**Comment by larskotthoff** [comment]
The latest version of ranger on CRAN is 0.6.0. Have you tried that?

**Comment by tjvananne** [comment]
That's odd, I installed it only a few days ago. I wonder how i got the 0.5.0. 

My apologies! I'll try that and let you all know. Thanks!

**Comment by tjvananne** [comment]
That did it. It works in 0.6.0, so still no reproducible example of original poster's issue (unless he too was using 0.5.0 through mlr and had factors in test that weren't found in train).

Loving ranger so far, thanks all!

**Comment by larskotthoff** [comment]
Thanks for checking, I'll close here.
